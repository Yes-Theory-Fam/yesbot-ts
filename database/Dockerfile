FROM postgres:13.0

RUN mkdir -p /docker-entrypoint-initdb.d
COPY setup.sql /docker-entrypoint-initdb.d/
