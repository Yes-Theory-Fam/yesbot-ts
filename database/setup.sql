CREATE ROLE yesbot WITH ENCRYPTED PASSWORD 'yesbot' LOGIN;
CREATE DATABASE yesbot OWNER yesbot;
