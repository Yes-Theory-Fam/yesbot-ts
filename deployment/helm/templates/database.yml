{{- if .Values.database.create -}}
kind: Service
apiVersion: v1
metadata:
  name: {{ .Values.database.service.name }}-{{ .Release.Name }}
  namespace: {{ .Release.Namespace }}
  labels: {{ .Values.database.service.labels | toYaml | nindent 4 }}
    release: {{ .Release.Name }}
spec:
  ports:
    - port: 5432
      targetPort: postgres
  selector: {{ .Values.database.service.labels | toYaml | nindent 4 }}

---

kind: Deployment
apiVersion: apps/v1
metadata:
  name: yesbot-database-deployment-{{ .Release.Name }}
  namespace: {{ .Release.Namespace }}
spec:
  selector:
    matchLabels: {{ .Values.database.service.labels | toYaml | nindent 6 }}
      release: {{ .Release.Name }}
  template:
    metadata:
      labels: {{ .Values.database.service.labels | toYaml | nindent 8 }}
        release: {{ .Release.Name }}
    spec:
      containers:
        - name: yesbot-database-deployment-{{ .Release.Name }}
          image: postgres:12.6
          ports:
            - name: postgres
              containerPort: 5432
          env:
            {{- range .Values.database.secret.secrets }}
            - name: {{ .name }}
              valueFrom:
                secretKeyRef:
                  name: {{ $.Values.database.secret.name }}-{{ $.Release.Name }}
                  key: {{ .name }}
            {{- end }}

---

{{- with .Values.database.secret }}
apiVersion: v1
kind: Secret
metadata:
  name: {{ .name }}-{{ $.Release.Name }}
  namespace: {{ $.Release.Namespace }}
data: {{- range $secret := .secrets }}
  {{ $secret.name }}: {{ $secret.value | toString | b64enc }}
  {{- end }}
{{- end }}

{{- end -}}
