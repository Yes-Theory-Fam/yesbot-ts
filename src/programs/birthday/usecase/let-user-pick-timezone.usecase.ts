import { getAllCountries, getCountry } from "countries-and-timezones";
import {
  CollectorFilter,
  GuildMember,
  Message,
  MessageEmbed,
  MessageReaction,
  User,
} from "discord.js";
import { createYesBotLogger } from "../../../log";
import { CountryRoleFinder } from "../../../utils/country-role-finder";

interface CountryWithRegion {
  country: string;
  region: string;
}

export class LetUserPickTimezoneUsecase {
  private readonly logger = createYesBotLogger(
    "birthday",
    LetUserPickTimezoneUsecase.name
  );

  async getUserTimezone(message: Message): Promise<string> {
    const countryRole = await this.fetchUserCountryRoles(message.member);

    const timezones = countryRole
      .map(this.timezonesFromRole)
      .reduce((prev, curr) => [...prev, ...curr], [])
      .filter((tz) => tz.includes("/"));

    if (timezones.length === 0) {
      throw new Error("No timezone found");
    }

    if (timezones.length > 20) {
      this.logger.error("User has too many available timezones: ", timezones);
      throw new Error("Too many available time zones");
    }

    const response = new MessageEmbed();
    response.setTitle("Pick your timezone");

    const regionIdentifierStart = 127462;
    const reactions: string[] = timezones.map((tz, i) => {
      const currentTime = new Date();
      const currentTimeString = `Current time: ${currentTime.toLocaleTimeString(
        "en-GB",
        { timeZone: tz }
      )}`;
      const identifier = String.fromCodePoint(regionIdentifierStart + i);
      response.addField(tz, `${identifier} - ${currentTimeString}`);
      return identifier;
    });

    const sentMessage = await message.channel.send({ embeds: [response] });
    for (const reaction of reactions) {
      try {
        await sentMessage.react(reaction);
      } catch (err) {
        this.logger.error("Error while adding timezones", err);
        // If we err here, it's probably because the user already selected an emoji.
        // Best to just skip adding more emojis.
        break;
      }
    }

    const filter: CollectorFilter<[MessageReaction, User]> = (
      reaction,
      user
    ) => {
      return (
        user.id === message.author.id && reactions.includes(reaction.emoji.name)
      );
    };

    let received;
    try {
      received = await sentMessage.awaitReactions({
        filter,
        max: 1,
        time: 60000,
        errors: ["time"],
      });
    } catch (err) {
      if (err.toString() === "[object Map]") {
        await sentMessage.delete();
        throw new Error("time expired");
      } else {
        throw err;
      }
    }

    const reaction = received.first();
    const selectedTz = timezones[reactions.indexOf(reaction.emoji.name)];

    await sentMessage.delete();

    return selectedTz;
  }

  private async fetchUserCountryRoles(
    user: GuildMember
  ): Promise<CountryWithRegion[]> {
    return user.roles.cache
      .filter((role) => CountryRoleFinder.isCountryRole(role.name, true))
      .map<CountryWithRegion>((role) => ({
        country: CountryRoleFinder.getCountryByRole(role.name, true),
        region: role.name.substring(
          role.name.indexOf("(") + 1,
          role.name.indexOf(")")
        ),
      }));
  }

  private timezonesFromRole(props: CountryWithRegion): readonly string[] {
    const { country, region } = props;
    // Edge cases
    switch (country) {
      case "USA":
        switch (region) {
          case "Southwest": {
            return ["America/Shiprock", "America/Phoenix"];
          }
          case "West": {
            return [
              "America/Atka",
              "America/Adak",
              "America/Anchorage",
              "America/Boise",
              "America/Denver",
              "America/Juneau",
              "America/Los_Angeles",
              "America/Metlakatla",
              "America/Nome",
              "America/Sitka",
              "America/Yakutat",
            ];
          }
          case "Midwest": {
            return [
              "America/Indianapolis",
              "America/Chicago",
              "America/Detroit",
              "America/Fort_Wayne",
              "America/Menominee",
            ];
          }
          case "Southeast": {
            return ["America/Louisville"];
          }
          case "Northeast": {
            return ["America/New_York"];
          }
          default: {
            // This is to slim down the extremely big list of US TZs.
            // Also, we remove any timezone that JS is unable to display.
            const usTZs = getCountry("US")
              .timezones.filter((tz) => tz.startsWith("America/"))
              .filter((tz) => tz.lastIndexOf("/") === tz.indexOf("/"));
            return usTZs
              .map((tz) => {
                try {
                  new Date().toLocaleTimeString("en-GB", { timeZone: tz });
                  return tz;
                } catch (e) {
                  return null;
                }
              })
              .filter((tz) => tz !== null);
          }
        }
      case "UK":
        return getCountry("GB").timezones;
      case "Mexico":
        return getCountry("MX").timezones;
      case "Australia": {
        return [
          "Australia/Perth",
          "Australia/Darwin",
          "Australia/Adelaide",
          "Australia/Brisbane",
          "Australia/Sydney",
        ];
      }
      case "Canada": {
        return getCountry("CA", { deprecated: true }).timezones.filter((tz) =>
          tz.startsWith("Canada/")
        );
      }
      case "Czech Republic": {
        return getCountry("CZ").timezones;
      }
      case "Brazil": {
        return [
          "Brazil/Acre",
          "Brazil/East",
          "Brazil/West",
          "America/Rio_Branco",
          "America/Sao_Paulo",
        ];
      }
      case "Argentina": {
        return ["America/Buenos_Aires"];
      }
      case "Chile": {
        return ["America/Santiago"];
      }
      case "UAE": {
        return getCountry("AE").timezones;
      }
      case "Russia": {
        return [
          "Europe/Kaliningrad",
          "Europe/Moscow",
          "Europe/Samara",
          "Asia/Yekaterinburg",
          "Asia/Omsk",
          "Asia/Novosibirsk",
          "Asia/Irkutsk",
          "Asia/Yakutsk",
          "Asia/Vladivostok",
          "Asia/Magadan",
          "Asia/Kamchatka",
        ];
      }
    }

    // let's find what tz's are available for this country
    let fixedCountry = country;
    if (country.startsWith("the "))
      fixedCountry = country.substring("the ".length);

    // REALLY
    const countries = getAllCountries();
    const correctCountry = Object.values(countries).find(
      (country) => country.name === fixedCountry
    );

    return correctCountry?.timezones ?? [];
  }
}
