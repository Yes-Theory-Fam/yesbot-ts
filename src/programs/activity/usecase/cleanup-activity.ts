import prisma from "../../../prisma";
import { Activity, Currency } from "@yes-theory-fam/database/client";
import { createYesBotLogger } from "../../../log";


export class CleanupActivity {
  private static cleanupActivityInstance: CleanupActivity;
  private static MS_PER_MINUTE = 60000;

  private delay =
    (!!process.env.ACTIVITY_TIME_DELAY
      ? Number.parseInt(process.env.ACTIVITY_TIME_DELAY)
      : 5) * CleanupActivity.MS_PER_MINUTE;

  private constructor() {}

  public static instance() {
    if (!CleanupActivity.cleanupActivityInstance) {
      CleanupActivity.cleanupActivityInstance = new CleanupActivity();
    }
    return CleanupActivity.cleanupActivityInstance;
  }

  public async findOldActivities() {
    const olderThen = new Date(Date.now() - this.delay);
    return await prisma.activity
      .findMany({
        take: 20,
        where: {
          updatedAt: {
            lt: olderThen,
          },
        },
      })
      .then((activities) => this.moveToCurrency(activities))
      .then(() => "finished");
  }

  private moveToCurrency(activities: Activity[]) {
    const data = activities.map((value) => {
      const activeTimeInMilliseconds =
        value.updatedAt.getTime() - value.createdAt.getTime();
      const activeMinutes: number = CleanupActivity.millisToMinutesAndSeconds(
        activeTimeInMilliseconds
      );
      return { userId: value.userId, activeMinutes };
    });

    const movedActivities = data.map((value) => {
      return prisma.currency
        .findUnique({ where: { userId: value.userId } })
        .then((user: Currency) => {
          if (!!user) {
            return CleanupActivity.updateCurrencyForUser(value, user);
          } else {
            return CleanupActivity.createCurrencyForUser(value);
          }
        });
    });

    return Promise.all(movedActivities)
      .then(async () => {
        return activities.map(async (value) => {
          return await prisma.activity.delete({
            where: { id: value.id },
          });
        });
      })
      .then(() => logger.info(`deleted activities: ${activities.length}`));
  }

  private static createCurrencyForUser(value: {
    activeMinutes: number;
    userId: any;
  }) {
    return prisma.currency.create({
      data: {
        userId: value.userId,
        ammount: value.activeMinutes,
      },
    });
  }

  private static updateCurrencyForUser(
    value: { activeMinutes: number; userId: any },
    user: Currency
  ) {
    return prisma.currency.update({
      where: {
        userId: value.userId,
      },
      data: {
        ammount: user.ammount + value.activeMinutes,
      },
    });
  }

  private static millisToMinutesAndSeconds(millis: number): number {
    return Math.floor(millis / 60000) + 1;
  }
}

const logger = createYesBotLogger("programs", CleanupActivity.name);
