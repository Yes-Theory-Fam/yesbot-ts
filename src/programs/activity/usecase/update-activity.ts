import { UpdateActivityType } from "../types/update-activity.type";
import prisma from "../../../prisma";
import { createYesBotLogger } from "../../../log";
import { Activity } from "@yes-theory-fam/database/client";

export class UpdateActivity {
  private static updateActivityInstance: UpdateActivity;
  private delay = new Date().setMinutes(
    Number.parseInt(process.env.ACTIVITY_TIME_DELAY) | 5
  );

  async handle(data: UpdateActivityType) {
    const date = new Date(Date.now() - this.delay);
    UpdateActivity.findActivity(data, date)
      .then((activity) => {
        if (!!activity) {
          return UpdateActivity.updateActivity(activity);
        } else {
          return UpdateActivity.createActivity(data);
        }
      })
      .then((data) => logger.debug(`activity created for ${data.userId}`));
  }

  private static createActivity(data: UpdateActivityType) {
    return prisma.activity.create({
      data: {
        userId: data.userId,
      },
    });
  }

  private static updateActivity(activity: Activity) {
    return prisma.activity.update({
      where: {
        id: activity.id,
      },
      data: {
        counter: activity.counter + 1,
      },
    });
  }

  private static findActivity(data: UpdateActivityType, date: Date) {
    return prisma.activity.findFirst({
      where: {
        AND: [
          { userId: data.userId },
          {
            updatedAt: {
              gte: date,
            },
          },
        ],
      },
    });
  }

  public static instance() {
    if (!UpdateActivity.updateActivityInstance) {
      UpdateActivity.updateActivityInstance = new UpdateActivity();
    }
    return UpdateActivity.updateActivityInstance;
  }
}

const logger = createYesBotLogger("programs", UpdateActivity.name);
