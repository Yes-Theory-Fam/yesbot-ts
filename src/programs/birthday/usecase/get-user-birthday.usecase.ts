import { Birthday } from "@yes-theory-fam/database/client";
import { utcToZonedTime } from "date-fns-tz";
import prisma from "../../../prisma";

export class GetUserBirthdayUseCase {
  async getUserBirthday(userId: string): Promise<Date | null> {
    const birthday = await prisma.birthday.findUnique({ where: { userId } });

    return birthday?.timezone
      ? utcToZonedTime(birthday.birthdate, birthday.timezone)
      : null;
  }
}
