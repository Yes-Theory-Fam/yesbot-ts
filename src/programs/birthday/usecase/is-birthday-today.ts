import {zonedTimeToUtc} from 'date-fns-tz';

import {Birthday} from '@yes-theory-fam/database/client';

export const isBirthdayToday = (birthday: Birthday, now = new Date(), targetYear: number = now.getFullYear()): boolean => {
  // Requirement: All birthdates are at time 00:00:00
  const birthdayInUtc = zonedTimeToUtc(birthday.birthdate, birthday.timezone);
  const thisTimezone = Intl.DateTimeFormat().resolvedOptions().timeZone;
  const nowInUtc = zonedTimeToUtc(now, thisTimezone);
  nowInUtc.setFullYear(targetYear);

  return nowInUtc.getFullYear() === birthdayInUtc.getFullYear() &&
    nowInUtc.getMonth() === birthdayInUtc.getMonth() &&
    nowInUtc.getDate() === birthdayInUtc.getDate();
};
