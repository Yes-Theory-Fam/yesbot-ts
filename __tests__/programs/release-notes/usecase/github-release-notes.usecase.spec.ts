import { GithubReleaseNotesUsecase } from "../../../../src/programs/release-notes/usecase/github-release-notes.usecase";
import {
  GithubReleaseRepositoryType,
  Node,
} from "../../../../src/programs/release-notes/types/github-release.types";
import { GithubReleaseNotesAccessor } from "../../../../src/programs/release-notes/acessor/github-release-notes.accessor";
import { Release } from "@yes-theory-fam/database/client";

jest.mock(
  "../../../../src/programs/release-notes/acessor/github-release-notes.accessor"
);

describe("GithubReleaseNotesUsecase", () => {
  let githubReleaseNotesAccessor: GithubReleaseNotesAccessor;
  let githubReleaseNotesUsecase: GithubReleaseNotesUsecase;

  beforeEach(() => {
    const getCommitMessagesResponse = {
      repository: {
        object: {
          history: {
            nodes: [
              {
                messageHeadline:
                  "chore(DOCKERFILE): docker image migrates th database by start",
              },
              {
                messageHeadline:
                  "chore(DOCKERFILE): docker image migrates th database by start",
              },
            ],
          },
        },
      },
    };
    githubReleaseNotesUsecase = new GithubReleaseNotesUsecase();

    githubReleaseNotesAccessor = new GithubReleaseNotesAccessor();
    githubReleaseNotesAccessor.getCommitMessages = jest.fn(() =>
      Promise.resolve(getCommitMessagesResponse)
    );
    githubReleaseNotesUsecase["githubReleaseNotesAccessor"] =
      githubReleaseNotesAccessor;
  });
  it("should sort Releases", () => {
    const input: GithubReleaseRepositoryType = {
      repository: {
        releases: {
          edges: [
            {
              node: {
                publishedAt: "2021-07-01T01:00:00Z",
              } as Node,
            },
            {
              node: {
                publishedAt: "2020-07-01T01:00:00Z",
              } as Node,
            },
          ],
        },
      },
    };

    const result = {
      oldRelease: { node: { publishedAt: "2020-07-01T01:00:00Z" } },
      newRelease: { node: { publishedAt: "2021-07-01T01:00:00Z" } },
    };

    expect(githubReleaseNotesUsecase["sortLatestReleases"](input)).toEqual(
      result
    );
  });

  it("should return git commit message", async () => {
    const release: Release = {
      id: "921332c4-64fb-46f2-8f8c-a6d6741934ed",
      releaseName: "test",
      releaseTime: "2020-07-01T01:00:00Z",
      releaseTag: "1.0.0",
      created: new Date("2020-07-01T01:00:00Z"),
    };

    githubReleaseNotesUsecase["getCommitMessages"](release, "2.0.0").then(
      (result) => {
        expect(result).toMatchSnapshot();
      }
    );
  });

  it("should create release notes", () => {
    const releaseNotes = `> * chore(DOCKERFILE): docker image migrates th database by start
> * chore(DOCKERFILE): docker image migrates th database by start\n`;
    const result =
      githubReleaseNotesUsecase["createReleaseNotes"](releaseNotes);
    expect(result).toMatchSnapshot();
  });
});
