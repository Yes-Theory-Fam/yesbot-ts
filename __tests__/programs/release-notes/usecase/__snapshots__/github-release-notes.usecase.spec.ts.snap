// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GithubReleaseNotesUsecase should create release notes 1`] = `
"Hello everyone :wave: I'm back after a short break :sleeping:
During this time I have been worked on a lot. Here is a small overview:

> * chore(DOCKERFILE): docker image migrates th database by start
> * chore(DOCKERFILE): docker image migrates th database by start

So now back to work :tools:     
"
`;

exports[`GithubReleaseNotesUsecase should return git commit message 1`] = `
"> * chore(DOCKERFILE): docker image migrates th database by start
> * chore(DOCKERFILE): docker image migrates th database by start
"
`;
